src/sparseml/pytorch/datasets/classification/_ffcv/__ffcv.py

```diff
@@ -0,0 +1,234 @@
+# Copyright (c) 2021 - present / Neuralmagic, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+from abc import ABC, abstractmethod
+from contextlib import contextmanager
+from pathlib import Path
+from typing import Callable, List, Optional
+
+import numpy as np
+import torch
+
+from ffcv.fields.basics import IntDecoder
+from ffcv.loader import OrderOption
+from sparseml.pytorch.utils import default_device
+
+
+try:
+    import ffcv
+    from ffcv import DatasetWriter, Loader
+    from ffcv.fields import IntField, RGBImageField
+    from ffcv.fields.rgb_image import (
+        CenterCropRGBImageDecoder,
+        RandomResizedCropRGBImageDecoder,
+    )
+    from ffcv.pipeline.operation import Operation
+    from ffcv.transforms import (
+        NormalizeImage,
+        RandomHorizontalFlip,
+        Squeeze,
+        ToDevice,
+        ToTensor,
+        ToTorchImage,
+    )
+except Exception as ffcv_error:
+    ffcv = None
+    ffcv_error = ffcv_error
+
+LOGGER = logging.getLogger(__name__)
+
+
+class FFCVDataset(ABC):
+    """
+    Abstract class for FFCV datasets.
+    """
+
+    def __init__(self, dataset, write_path: str = "data/ffcv/", *args, **kwargs):
+        """
+        Initialize the dataset.
+        """
+        self.dataset = dataset
+        self.write_path = write_path
+        Path(self.write_path).parent.mkdir(parents=True, exist_ok=True)
+
+    @abstractmethod
+    def _write(self, *args, **kwargs):
+        """
+        Write the dataset to disk.
+        """
+        pass
+
+    @abstractmethod
+    def switched_off_transform(self, *args, **kwargs):
+        """
+        Switch off the transform.
+        """
+        pass
+
+
+class ImageFolderFFCV(FFCVDataset):
```

Comment:
considering restructuring instead of a wrapper class to make it a co-parent class for our datasets - this will give each dataset the potential to override transforms used